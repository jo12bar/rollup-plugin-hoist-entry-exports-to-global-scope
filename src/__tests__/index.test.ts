import { rollup, RollupOutput, OutputChunk, OutputAsset } from 'rollup';
import hoistExports from '../index';
import * as path from 'path';

function isRollupChunk(x: OutputChunk | OutputAsset): x is OutputChunk {
  return x.type === 'chunk';
}

async function generateUmdAndIifeBundles(
  relativePath: string,
  bundleName: string
): Promise<{ umd: RollupOutput; iife: RollupOutput }> {
  const bundle = await rollup({
    input: path.join(__dirname, relativePath),
    plugins: [hoistExports(bundleName)],
  });

  const umd = await bundle.generate({ name: bundleName, format: 'umd' });
  const iife = await bundle.generate({ name: bundleName, format: 'iife' });

  return { umd, iife };
}

test('generates bindings for variable-only file', async () => {
  const { umd, iife } = await generateUmdAndIifeBundles(
    './fixtures/just-variables.js',
    'JustVariables'
  );

  expect(
    umd.output.filter(isRollupChunk).map((chunk) => chunk.code)
  ).toMatchSnapshot();
  expect(
    iife.output.filter(isRollupChunk).map((chunk) => chunk.code)
  ).toMatchSnapshot();
});

test('generates bindings for function-only file', async () => {
  const { umd, iife } = await generateUmdAndIifeBundles(
    './fixtures/just-functions.js',
    'JustFunctions'
  );

  expect(
    umd.output.filter(isRollupChunk).map((chunk) => chunk.code)
  ).toMatchSnapshot();
  expect(
    iife.output.filter(isRollupChunk).map((chunk) => chunk.code)
  ).toMatchSnapshot();
});

test('generates bindings for class-only file', async () => {
  const { umd, iife } = await generateUmdAndIifeBundles(
    './fixtures/just-classes.js',
    'JustClasses'
  );

  expect(
    umd.output.filter(isRollupChunk).map((chunk) => chunk.code)
  ).toMatchSnapshot();
  expect(
    iife.output.filter(isRollupChunk).map((chunk) => chunk.code)
  ).toMatchSnapshot();
});

test('generates bindings for files that contain imports', async () => {
  const { umd, iife } = await generateUmdAndIifeBundles(
    './fixtures/importing-things.js',
    'ImportingThings'
  );

  expect(
    umd.output.filter(isRollupChunk).map((chunk) => chunk.code)
  ).toMatchSnapshot();
  expect(
    iife.output.filter(isRollupChunk).map((chunk) => chunk.code)
  ).toMatchSnapshot();
});
