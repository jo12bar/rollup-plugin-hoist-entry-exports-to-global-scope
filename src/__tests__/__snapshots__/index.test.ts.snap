// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates bindings for class-only file 1`] = `
Array [
  "(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.JustClasses = {}));
}(this, (function (exports) { 'use strict';

  class TestClass1 {
    constructor() {
      this.foo = 'bar';
    }
  }

  class TestClass2 extends TestClass1 {
    constructor() {
      super();
      this.spam = this.foo + ' and eggs';
    }
  }

  exports.TestClass1 = TestClass1;
  exports.TestClass2 = TestClass2;

  Object.defineProperty(exports, '__esModule', { value: true });

})));

var TestClass1=JustClasses.TestClass1;var TestClass2=JustClasses.TestClass2;
",
]
`;

exports[`generates bindings for class-only file 2`] = `
Array [
  "var JustClasses = (function (exports) {
  'use strict';

  class TestClass1 {
    constructor() {
      this.foo = 'bar';
    }
  }

  class TestClass2 extends TestClass1 {
    constructor() {
      super();
      this.spam = this.foo + ' and eggs';
    }
  }

  exports.TestClass1 = TestClass1;
  exports.TestClass2 = TestClass2;

  Object.defineProperty(exports, '__esModule', { value: true });

  return exports;

}({}));

var TestClass1=JustClasses.TestClass1;var TestClass2=JustClasses.TestClass2;
",
]
`;

exports[`generates bindings for files that contain imports 1`] = `
Array [
  "(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ImportingThings = {}));
}(this, (function (exports) { 'use strict';

  class TestClass1 {
    constructor() {
      this.foo = 'bar';
    }
  }

  class TestClass2 extends TestClass1 {
    constructor() {
      super();
      this.spam = this.foo + ' and eggs';
    }
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  function identity(x) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return x;
  }

  let spam = ['eggs', 'eggs', 'eggs'];

  class CompositeClass extends TestClass2 {
    constructor() {
      super();
      /* eslint-disable @typescript-eslint/restrict-plus-operands */
      this.yeet =
        this.spam + this.foo + identity(' yeet') + ' ' + spam.join(' ');
      /* eslint-enable @typescript-eslint/restrict-plus-operands */
    }
  }

  exports.CompositeClass = CompositeClass;

  Object.defineProperty(exports, '__esModule', { value: true });

})));

var CompositeClass=ImportingThings.CompositeClass;
",
]
`;

exports[`generates bindings for files that contain imports 2`] = `
Array [
  "var ImportingThings = (function (exports) {
  'use strict';

  class TestClass1 {
    constructor() {
      this.foo = 'bar';
    }
  }

  class TestClass2 extends TestClass1 {
    constructor() {
      super();
      this.spam = this.foo + ' and eggs';
    }
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  function identity(x) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return x;
  }

  let spam = ['eggs', 'eggs', 'eggs'];

  class CompositeClass extends TestClass2 {
    constructor() {
      super();
      /* eslint-disable @typescript-eslint/restrict-plus-operands */
      this.yeet =
        this.spam + this.foo + identity(' yeet') + ' ' + spam.join(' ');
      /* eslint-enable @typescript-eslint/restrict-plus-operands */
    }
  }

  exports.CompositeClass = CompositeClass;

  Object.defineProperty(exports, '__esModule', { value: true });

  return exports;

}({}));

var CompositeClass=ImportingThings.CompositeClass;
",
]
`;

exports[`generates bindings for function-only file 1`] = `
Array [
  "(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.JustFunctions = {}));
}(this, (function (exports) { 'use strict';

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  function fn1() {
    return true;
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  function fn2() {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return this;
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  function identity(x) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return x;
  }

  exports.fn1 = fn1;
  exports.fn2 = fn2;
  exports.identity = identity;

  Object.defineProperty(exports, '__esModule', { value: true });

})));

function fn1(){return JustFunctions.fn1.apply(null,arguments);}function fn2(){return JustFunctions.fn2.apply(null,arguments);}function identity(){return JustFunctions.identity.apply(null,arguments);}
",
]
`;

exports[`generates bindings for function-only file 2`] = `
Array [
  "var JustFunctions = (function (exports) {
  'use strict';

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  function fn1() {
    return true;
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  function fn2() {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return this;
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  function identity(x) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return x;
  }

  exports.fn1 = fn1;
  exports.fn2 = fn2;
  exports.identity = identity;

  Object.defineProperty(exports, '__esModule', { value: true });

  return exports;

}({}));

function fn1(){return JustFunctions.fn1.apply(null,arguments);}function fn2(){return JustFunctions.fn2.apply(null,arguments);}function identity(){return JustFunctions.identity.apply(null,arguments);}
",
]
`;

exports[`generates bindings for variable-only file 1`] = `
Array [
  "(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
	typeof define === 'function' && define.amd ? define(['exports'], factory) :
	(global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.JustVariables = {}));
}(this, (function (exports) { 'use strict';

	const foo = 'bar';
	let spam = ['eggs', 'eggs', 'eggs'];
	var context = hiddenContext;
	var hiddenContext = -42;

	exports.context = context;
	exports.foo = foo;
	exports.spam = spam;

	Object.defineProperty(exports, '__esModule', { value: true });

})));

var foo=JustVariables.foo;var spam=JustVariables.spam;var context=JustVariables.context;
",
]
`;

exports[`generates bindings for variable-only file 2`] = `
Array [
  "var JustVariables = (function (exports) {
	'use strict';

	const foo = 'bar';
	let spam = ['eggs', 'eggs', 'eggs'];
	var context = hiddenContext;
	var hiddenContext = -42;

	exports.context = context;
	exports.foo = foo;
	exports.spam = spam;

	Object.defineProperty(exports, '__esModule', { value: true });

	return exports;

}({}));

var foo=JustVariables.foo;var spam=JustVariables.spam;var context=JustVariables.context;
",
]
`;
